File: Games/RockPaperScissors/src/main/java/jokatu/game/games/rockpaperscissors/RockPaperScissors.java
package jokatu.game.games.rockpaperscissors;

import jokatu.game.result.Result;

import static jokatu.game.result.Result.*;

/**
 * An enum that contains entries for rock, paper and scissors.
 */
public enum RockPaperScissors {
	ROCK {
		@Override
		public Result resultAgainst(RockPaperScissors other) {
			switch (other) {
				case ROCK:
					return DRAW;
				case PAPER:
					return LOSE;
				case SCISSORS:
					return WIN;
				default:
					return DRAW;
			}
		}
	},
	PAPER {
		@Override
		public Result resultAgainst(RockPaperScissors other) {
			switch (other) {
				case ROCK:
					return WIN;
				case PAPER:
					return DRAW;
				case SCISSORS:
					return LOSE;
				default:
					return DRAW;
			}
		}
	},
	SCISSORS {
		@Override
		public Result resultAgainst(RockPaperScissors other) {
			switch (other) {
				case ROCK:
					return LOSE;
				case PAPER:
					return WIN;
				case SCISSORS:
					return DRAW;
				default:
					return DRAW;
			}
		}
	};

	public abstract Result resultAgainst(final RockPaperScissors other);
}

File: Games/RockPaperScissors/src/main/kotlin/jokatu/game/games/rockpaperscissors/game/RockPaperScissors.kt
package jokatu.game.games.rockpaperscissors.game

import jokatu.game.result.Result

import jokatu.game.result.Result.*

/**
 * An enum that contains entries for rock, paper and scissors.
 */
enum class RockPaperScissors {
	ROCK {
		override fun resultAgainst(other: RockPaperScissors): Result {
			return result(other, DRAW, LOSE, WIN)
		}
	},
	PAPER {
		override fun resultAgainst(other: RockPaperScissors): Result {
			return result(other, WIN, DRAW, LOSE)
		}
	},
	SCISSORS {
		override fun resultAgainst(other: RockPaperScissors): Result {
			return result(other, LOSE, WIN, DRAW)
		}
	};

	fun result(other: RockPaperScissors, againstRock: Result, againstPaper: Result, againstScissors: Result): Result {
		return when (other) {
			ROCK -> againstRock
			PAPER -> againstPaper
			SCISSORS -> againstScissors
		}
	}

	abstract fun resultAgainst(other: RockPaperScissors): Result
}
